content/posts/windows-containers.md

```diff
@@ -0,0 +1,231 @@
+  
+
+
+    [![](/img/windows-containers/error.png)](/img/windows-containers/error.png)*Scheduling error when not using nodeSelectors. Debug using ```kubectl describe podname```*
+  
+
+- Jenkinsfile:
+
+```
+pipeline {
+  agent none
+  options { 
+    buildDiscarder(logRotator(numToKeepStr: '2'))
+    skipDefaultCheckout true
+  }
+  stages {
+    stage('Test-linux') {
+      agent {
+        kubernetes {
+          label 'nodejs-pod'
+          yamlFile 'linux/nodejs-pod.yaml'
+        }
+      }
+      steps {
+        checkout scm
+        container('nodejs') {
+          echo 'Hello World!'   
+          sh 'node --version'
+        }
+      }
+    }
+    stage('Test-windows') {
+      agent {
+        kubernetes {
+	  label 'windows-pod'
+          yamlFile 'windows/dotnet-pod.yaml'
+        }
+      }
+      steps {
+        bat 'dir'
+        container(name:'windows-dotnet'){
+          bat 'dotnet -h'
+      } 
+     }
+    }
+  }
+}
+```
+
+
+- This is a declarative pipeline using agent ```none```([none](https://jenkins.io/doc/book/pipeline/syntax/#agent) info) so that we can specify agents per stage. ```yamlFile``` is used to read the pod template from a file location which is also in this repo. You can either define the pod template in another file, in the same [Jenkinsfile](https://github.com/jenkinsci/kubernetes-plugin/blob/master/examples/declarative-multiple-containers.groovy) or in the Jenkins Configuration page. More info about the syntax used in this pipeline can be found [here](https://jenkins.io/doc/book/pipeline/syntax/). 
+
+
+- The ```node --version``` command is executed inside the ```container``` step otherwise it will get executed inside the jnlp (Linux) container and fail as it doesn't have node installed. Similarly ```bat 'dir ``` gets executed in the jnlp (Windows) container and ```bat 'dotnet -h' ``` is executed inside the dotnet container.
+
+
+
+- [Here](https://github.com/jenkinsci/kubernetes-plugin/blob/kubernetes-1.21.0/examples/windows.groovy) is another example on how to use a Windows container using a scripted pipeline and tested in EKS.
+
+- [Windows Containers on the Kubernetes Podcast](https://open.spotify.com/episode/0XXYzjBEj12S39rwcobJ70?si=J-tEvxR8S_qsmJcd-FGn5A)
+
+ [![](/img/windows-containers/node-pools-distribution.png)](/img/windows-containers/node-pools-distribution.png)*This is a high-level diagram of the Kubernetes cluster and containers.*
+
```

Comment:
I think the post would close more strongly with a bit more of a conclusion - maybe just a sentence or two recapping what was walked through before suggesting these additional resources.